Update buildbot_steps lint to handle env variables

aneeshusa opened the pull request:
https://github.com/servo/saltfs/pull/687 added support
for specifying environment variables in `buildbot_steps.yml`.
Update the servo-tidy buildbot_steps.yml linter to reflect this.

Use the voluptuous Python library (BSD 3-clause license) for validation
in lieu of a much larger hand-written implementation.

Extracted out of #17171. Helps with servo/saltfs#770.

<!-- Thank you for contributing to Servo! Please replace each `[ ]` by `[X]` when the step is complete, and replace `__` with appropriate data: -->
- [ ] `./mach build -d` does not report any errors
- [x] `./mach test-tidy` does not report any errors
- [ ] These changes fix #__ (github issue number if applicable).

<!-- Either: -->
- [x] There are tests for these changes OR
- [ ] These changes do not require tests because they change the tests

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/servo/servo/19720)
<!-- Reviewable:end -->

aneeshusa committed:
Update buildbot_steps lint to handle env variables

https://github.com/servo/saltfs/pull/687 added support
for specifying environment variables in `buildbot_steps.yml`.
Update the servo-tidy buildbot_steps.yml linter to reflect this.

Use the voluptuous Python library (BSD 3-clause license) for validation
in lieu of a much larger hand-written implementation.
Update the tidy self tests to take into account the new error messages.

highfive commented:
Heads up! This PR modifies the following files:
 * @wafflespeanut: python/requirements.txt, python/tidy/servo_tidy/tidy.py
 * @edunham: python/tidy/servo_tidy/tidy.py

aneeshusa commented:
cc @jdm

aneeshusa committed:
Deny warnings in rust code globally in CI

Use the new environment variable support for `buildbot_steps.yml`
to do this globally instead of having to patch every call,
which is error prone.

Set this variable globally in .travis.yml and appveyor.yml as well.

aneeshusa commented:
I've added a second commit which redoes #19612 (denying warnings in CI) with this new support as an example of its usefulness.

aneeshusa commented:
Fixed tidy's self test and updated `.travis.yml` and `appveyor.yml` as well.

jdm commented:
This is a nice improvement. Thanks!
@bors-servo r+

bors-servo commented:
:pushpin: Commit 94febf4 has been approved by `jdm`

<!-- @bors-servo r=jdm 94febf43d6a989e0859e0c5454a3e4e7c317cacc -->

bors-servo commented:
:hourglass: Testing commit 94febf43d6a989e0859e0c5454a3e4e7c317cacc with merge 5e110ee0b2cf98d4e731ba437cad6a02d884724b...

bors-servo commented:
:broken_heart: Test failed - [linux-dev](http://build.servo.org/builders/linux-dev/builds/10601)

jdm commented on etc/ci/buildbot_steps.yml:
```diff
@@ -32,11 +35,11 @@ mac-rel-wpt4:
-  - env PKG_CONFIG_PATH=/usr/local/opt/zlib/lib/pkgconfig RUSTFLAGS=-Dwarnings ./mach build-cef
-  - env RUSTFLAGS=-Dwarnings ./mach build-geckolib
+  - env PKG_CONFIG_PATH=/usr/local/opt/zlib/lib/pkgconfig ./mach build-cef
+  - env ./mach build-geckolib
```
We'll need to remove all of the now-unnecessary env prefixes.

SimonSapin commented:
@aneeshusa I’ve pushed a fixup commit to your branch

@bors-servo try

bors-servo commented:
:hourglass: Trying commit be1f769ada8180b82fa11f2bcf38fe9478f02092 with merge 440275f8090cbd1a696fb0eb560656485c46f515...

bors-servo commented:
:broken_heart: Test failed - [linux-dev](http://build.servo.org/builders/linux-dev/builds/10619)

SimonSapin commented:
```
env CC=gcc-5 CXX=g++-5 "" ./mach build-cef
[…]
env: : No such file or directory
```

That `""` comes form having two consecutive spaces in `etc/ci/buildbot_steps.yml`, with the empty string between them being interpreted as a separate argument. Some bit of Python in servo/saltfs probably need to use `foo.split()` instead of `foo.split(" ")`, or something like that. I’ve removed the extra space.

jdm commented:
@bors-servo r+

bors-servo commented:
:pushpin: Commit e9eaadc has been approved by `jdm`

<!-- @bors-servo r=jdm e9eaadc913c62ee62d90bcacdf61436034543a54 -->

bors-servo commented:
:hourglass: Testing commit e9eaadc913c62ee62d90bcacdf61436034543a54 with merge e8015fd0f31aa8b6f9fde95e1a6a343cc50de671...

bors-servo commented:
:sunny: Test successful - [android](http://build.servo.org/builders/android/builds/10332), [arm32](http://build.servo.org/builders/arm32/builds/10568), [arm64](http://build.servo.org/builders/arm64/builds/10594), [linux-dev](http://build.servo.org/builders/linux-dev/builds/10620), [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/7480), [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/7495), [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/10697), [mac-rel-css1](http://build.servo.org/builders/mac-rel-css1/builds/3106), [mac-rel-css2](http://build.servo.org/builders/mac-rel-css2/builds/3125), [mac-rel-wpt1](http://build.servo.org/builders/mac-rel-wpt1/builds/7375), [mac-rel-wpt2](http://build.servo.org/builders/mac-rel-wpt2/builds/7435), [mac-rel-wpt3](http://build.servo.org/builders/mac-rel-wpt3/builds/3195), [mac-rel-wpt4](http://build.servo.org/builders/mac-rel-wpt4/builds/3144), [windows-msvc-dev](http://build.servo.org/builders/windows-msvc-dev/builds/5995)
Approved by: jdm
Pushing e8015fd0f31aa8b6f9fde95e1a6a343cc50de671 to master...

SimonSapin commented:
@bors-servo retry

bors-servo commented:
:hourglass: Testing commit e9eaadc913c62ee62d90bcacdf61436034543a54 with merge 6a6dda526961b7bb7b856e8310ffc165d8bd39aa...

bors-servo commented:
:sunny: Test successful - [android](http://build.servo.org/builders/android/builds/10333), [arm32](http://build.servo.org/builders/arm32/builds/10569), [arm64](http://build.servo.org/builders/arm64/builds/10595), [linux-dev](http://build.servo.org/builders/linux-dev/builds/10621), [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/7481), [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/7497), [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/10699), [mac-rel-css1](http://build.servo.org/builders/mac-rel-css1/builds/3107), [mac-rel-css2](http://build.servo.org/builders/mac-rel-css2/builds/3126), [mac-rel-wpt1](http://build.servo.org/builders/mac-rel-wpt1/builds/7376), [mac-rel-wpt2](http://build.servo.org/builders/mac-rel-wpt2/builds/7436), [mac-rel-wpt3](http://build.servo.org/builders/mac-rel-wpt3/builds/3196), [mac-rel-wpt4](http://build.servo.org/builders/mac-rel-wpt4/builds/3145), [windows-msvc-dev](http://build.servo.org/builders/windows-msvc-dev/builds/5996)
Approved by: jdm
Pushing 6a6dda526961b7bb7b856e8310ffc165d8bd39aa to master...
